Extra docs for the pushdown feature

leskin-in opened the pull request:
This PR proposes extra documentation for #4968.

Update GPDB external tables documentation: add a reference and a short description of a GUC parameter `gp_external_enable_filter_pushdown`.

Describe what data types and operators are supported by pushdown implementation of the GPDB PXF extension.

kongyew commented:
@leskin-in  Thanks for your suggestions.  @dyozie  Please help to review the proposal.

dyozie commented on gpdb-doc/dita/admin_guide/external/g-external-tables.xml:
```diff
@@ -58,18 +58,30 @@
+               <li id="du210102">Reading external table data in parallel 
+               from multiple Greenplum database segment instances to optimize large load operations;</li>
+               <li id="du210103">Filter pushdown (if a query contains WHERE clause, 
+               it may be passed to external data source). See <xref href="../../../ref_guide/config_params/guc-list.xml#gp_external_enable_filter_pushdown">. 
```
Small edit:  "passed to external" -> "passed to the external"

dyozie commented on gpdb-doc/dita/admin_guide/external/g-external-tables.xml:
```diff
@@ -58,18 +58,30 @@
+               from multiple Greenplum database segment instances to optimize large load operations;</li>
+               <li id="du210103">Filter pushdown (if a query contains WHERE clause, 
+               it may be passed to external data source). See <xref href="../../../ref_guide/config_params/guc-list.xml#gp_external_enable_filter_pushdown">. 
+               Note that this feature may be not supported by some external data sources.</li>
```
We should call out which ones are known to support/not support push-down here.

dyozie commented on gpdb-doc/markdown/pxf/using_pxf.html.md.erb:
```diff
@@ -169,3 +169,26 @@ Greenplum Database passes the parameters in the `LOCATION` string as headers to
+
+To enable this feature (not only for PXF, but for all external data accessors in Greenplum database), set the GUC parameter `gp_external_enable_filter_pushdown` to `on` (default value is `off`).
+
+Not all external data sources support this feature. If such source is accessed, the query will be executed without pushdown. Also, not all data types and operators are supported.
```
Small edit:  "such source" -> "such a source"

lisakowen commented:
general comments:
- a few pages of the docs specifically state that PXF does not support pushdown, we should remove these statements:
    - dita/admin_guide/external/pxf-overview.xml  
    - markdown/pxf/hbase_pxf.html.md.erb
- can we identify which connectors and/or profiles actually support filter pushdown?  i.e. do all profiles provided by the Hive Connector support pushdown, or only specific profiles?

lisakowen commented on gpdb-doc/dita/admin_guide/external/g-external-tables.xml:
```diff
@@ -58,18 +58,30 @@
+               from multiple Greenplum database segment instances to optimize large load operations;</li>
+               <li id="du210103">Filter pushdown (if a query contains WHERE clause, 
+               it may be passed to external data source). See <xref href="../../../ref_guide/config_params/guc-list.xml#gp_external_enable_filter_pushdown">. 
+               Note that this feature may be not supported by some external data sources.</li>
```
when the user creates an external table, they provide a LOCATION clause specifying a specific protocol (pxf, gphdfs, gpfdist, s3).  it seems like there are two levels of support required:
- the protocol must support/implement filter pushdown, and 
- for a protocol that supports filter pushdown (currently only pxf), the external data source accessed by the protocol must also support it

is the above correct?  if so, do we want to specifically state that the protocol you use to create the external table must support filter pushdown, and that PXF is the only greenplum database protocol that supports it?

leskin-in replied:
Yes, you are correct. Actually, the external database (or other source of data) must also support the pushdown feature, as it is provided by the PXF connector being used to access it (consider the database that is accessed via JDBC (implemented in PXF), and that does not accept or does not process `WHERE` expressions in `SELECT`).

I have added a complete list of levels that must support pushdown, see https://github.com/greenplum-db/gpdb/pull/5193/commits/c1d9b4eee4aa723f6a220d49baaa9d4654e70585#diff-f3c93cd71c69b9b765844e0cdee2cfb1R199.

dyozie commented:
Thanks for the updates @leskin-in   

@kongyew - can you answer the questions about which of the current connectors/profiles actually support pushdown?

dyozie commented:
Thanks again, Ivan.  I do notice that the Hive and HiveVectorizedORC profiles support pushdown in Pivotal HDB.  I'm not sure if the recent code PR means they should also be supported in GPDB now.  If it's just a matter of us not having tested this in GPDB yet, that's a good reason to keep the non-support statement.

I also just realized that @kongyew is out for a while.  @denalex - can you clarify the above question about Hive/HiveVectorizedORC and predicate pushdown?

leskin-in commented:
@dyozie , @kongyew ,
As far as I understand, [Hive](https://github.com/apache/incubator-hawq/blob/e082d3adbb89a9e5ccb341da324ea9514cff5a35/pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveFilterBuilder.java#L40) and [HBase](https://github.com/apache/incubator-hawq/blob/e082d3adbb89a9e5ccb341da324ea9514cff5a35/pxf/pxf-hbase/src/main/java/org/apache/hawq/pxf/plugins/hbase/HBaseFilterBuilder.java#L58) PXF connectors also support pushdown.

leskin-in commented:
@dyozie ,
Taking your words and the PXF code into consideration, I have added the statement that Hive and HBase do support pushdown. However, me and @kapustor have not tested that. It would be great if anyone could confirm that those connectors really support the pushdown.

I have also removed the references to JDBC and Ignite connectors. Actually, it is not clear to me what tests should the PXF connector pass so that it becomes an officially supported one.

dyozie commented:
Thanks, Ivan - I'm going to go ahead and merge this in order to do some copyedits on the material.  I'll still double-check with Kong to make sure he's ok with the Hive and HBase statements, but it seems likely if the code is the same as in HAWQ.
